docs/design/2021-04-26-lock-view.md

```diff
@@ -0,0 +1,196 @@
+* [Unresolved Questions](#unresolved-questions)
+
+## Introduction
+
+This document describes the design of the feature Lock View, which provides tools to analyze problems about transaction's lock waiting, lock contentions and deadlocks.
+
+## Motivation or Background
+
+Currently, it's very hard to analyze lock contentions and deadlocks for transactions. One may need to enable general log, try to reproduce the problem, and try to analyze the log to find the cause, which is very difficult and inconvenient. We also found that this way of analyzing is not feasible in some scenarios. It's highly required to provide some better approach to analyze these kinds of problems.
+
+## Detailed Design
+
+Several tables will be provided in `information_schema`. Some tables has both local version (fetches data on the current TiDB node) and global version (fetches data among the whole cluster), and the global version's table name has the `"CLUSTER_"` prefix.
+
+### Table `(CLUSTER_)TIDB_TRX`
+
+| Field      | Type       | Comment |
+|------------|------------|---------|
+|`TRX_ID`    | `unsigned bigint`   | The transaction ID (aka. start ts) |
+|`TRX_STARTED`|`time`| Human readable start time of the transaction |
+|`DIGEST`|`text`| The digest of the current executing SQL statement |
+|`SQLS` | `text` | A list of all executed SQL statements' digests |
+|`STATE`| `enum('Running', 'Lock waiting', 'Committing', 'RollingBack')`| The state of the transaction |
+| `WAITING_START_TIME` | `time` | The elapsed time since the start of the current lock waiting (if any) |
+| `SCOPE` | `enum('Global', 'Local')` | The scope of the transaction |
+| `ISOLATION_LEVEL` | `enum('RR', 'RC')` | |
+| `AUTOCOMMIT` | `bool` |  |
+| `SESSION_ID` | `unsigned bigint` | | 
+| `USER` | `varchar` | |
+| `DB` | `varchar` | |
+| `SET_COUNT` | `int` | Modified keys of the current transaction |
+| `LOCKED_COUNT` | `int` | Locked keys of the current transaction |
+| `MEM_BUFFER_SIZE` | `int` | Size occupied by the transaction's membuffer |
+
+* Life span of rows:
+    * Create on first writing or locking operation in a transaction
+    * Remove after the transaction is done
+* Collecting, storing and querying:
+    * All these information can be collected on TiDB side. Since the amount of concurrent transactions won't be too large, and it doesn't need to be persisted, so it's ok to implement it as a memory table. For querying among the cluster, just register the table under `infoschema/cluster.go` and write the global table name with the local one.
+    * As the simplest way of implementing, most information can be passed with a similar way like `ProcessInfo`, or even directly passed via the `ProcessInfo` struct.
+* Permission:
+    * `PROCESS` privilege is needed to access the full content of this table. For users without `PROCESS` permission, only transactions started by the current user will be shown, and others will be filtered out, which is similar to the `processlist` table.
```

Comment:
Does mysql has the similar access requirement for this information?

Reply:
I didn't find any complete explanation. Here are some pieces of information I've found.
* The `INNODB_TRX` table of MySQL 5.7 requires `PROCESS` privilege ( [ref](https://dev.mysql.com/doc/refman/5.7/en/information-schema-innodb-trx-table.html) )
* The `INNODB_LOCKS` table of MySQL 5.7 requires `PROCESS` privilege ( [ref](https://dev.mysql.com/doc/refman/5.7/en/information-schema-innodb-locks-table.html) )
* The `INNODB_LOCK_WAITS` table of MySQL 5.7 requires `PROCESS` privilege ( [ref](https://dev.mysql.com/doc/refman/5.7/en/information-schema-innodb-lock-waits-table.html) )
* The `SHOW ENGINE` statement of MySQL 5.7, which provides a tool to analyze deadlocks, requires `PROCESS` privilege. ( [ref](https://dev.mysql.com/doc/refman/5.7/en/show-engine.html) )
* The `DATA_LOCKS` and `DATA_LOCK_WATIS` table of MySQL 8.0 has different permission requirement, which I think I didn't understand. ( [ref](https://dev.mysql.com/doc/refman/8.0/en/information-schema-innodb-locks-table.html) )